Physical constants epoch not working when using Kassiopeia as an external library

> Regarding the selection of the physical constants epoch, we do not presently have a known way of doing the flag selection from P8 through the ```#ifndef..#endif``` logic in prebuilt images (the combined build crashes, even with cmake flags specified).  As a temporary workaround we have hard-wired the 2006 constants in our local fork of Kass.  Would it be possible to somehow alter the ```#ifndef..#endif``` logic, or remove it and hard-wire the 2021 constants?  We can then update to the 2021 constants on the P8 side, although possibly not quite instantaneously.  Alternatively, we can explore the cmake flag usage further, which might also be useful.  Maybe that should have worked in the first place.

_Originally posted by @pslocum in https://github.com/KATRIN-Experiment/Kassiopeia/issues/71#issuecomment-1714379658_

Below is the first indication of an error during our build with the KConst option, even with
`-D KConst_REFERENCE_EPOCH:STRING=2006` in our P8 CMakeLists.txt file.  The error looks as though a value for the flag has not been specified.  It is possible that there is a more appropriate way to specify the option.  But our present workaround is to remove the option, by replacing the file `KConst.h` with `KConst_2006.h` .  


/usr/local/p8/kassiopeia/vDebug4.0.0/include/KConst.h:462:2: error: #error "Unsupported value for KConst_REFERENCE_EPOCH."
  462 | #error "Unsupported value for KConst_REFERENCE_EPOCH."
      |  ^~~~~
In file included from /usr/local/p8/kassiopeia/vDebug4.0.0/include/KTwoVector.hh:4,
                 from /usr/local/p8/kassiopeia/vDebug4.0.0/include/KThreeVector.hh:6,
                 from /usr/local/p8/kassiopeia/vDebug4.0.0/include/KThreeMatrix.hh:4,
                 from /usr/local/p8/kassiopeia/vDebug4.0.0/include/KSElectricField.h:6,
                 from /usr/local/p8/kassiopeia/vDebug4.0.0/include/KSParticle.h:4,
                 from /usr/local/p8/kassiopeia/vDebug4.0.0/include/KSSpaceInteraction.h:5,
                 from /tmp_loc_source/Source/Kassiopeia/LMCFieldCalculator.hh:11,
                 from /tmp_loc_source/Source/Generators/LMCCavitySignalGenerator.hh:20,
                 from /tmp_loc_source/Source/Generators/LMCCavitySignalGenerator.cc:8:
/usr/local/p8/kassiopeia/vDebug4.0.0/include/KConst.h:462:2: error: #error "Unsupported value for KConst_REFERENCE_EPOCH."
  462 | #error "Unsupported value for KConst_REFERENCE_EPOCH."
      |  ^~~~~

--- a/Kommon/Base/CMakeLists.txt
+++ b/Kommon/Base/CMakeLists.txt
@@ -41,8 +41,8 @@ set(KOMMON_BASE_HEADER_FILES
     Utility/KAlgorithm.h
     Utility/KException.h
     Utility/KCast.h
-    Utility/KConst.h
     Utility/KConst_2006.h
+    Utility/KConst_2021.h
     Utility/KConsoleMuter.h
     Utility/KHash.h
     Utility/OstreamJoiner.h
@@ -193,8 +193,10 @@ elseif(KConst_REFERENCE_EPOCH LESS 2021)
     "of some calculations. To restore *default* behavior, set KConst_REFERENCE_EPOCH=2021.")
 endif()
 
-# Needed for physical constants defined in Utility/KConst.h
-target_compile_definitions(KommonBase PUBLIC KConst_REFERENCE_EPOCH=${KConst_REFERENCE_EPOCH})
+# Bake KConst_REFERENCE_EPOCH into KConst.h
+configure_file( Utility/KConst.h.in ${CMAKE_CURRENT_BINARY_DIR}/Utility/KConst.h @ONLY )
+list( APPEND KOMMON_BASE_HEADER_FILES ${CMAKE_CURRENT_BINARY_DIR}/Utility/KConst.h )
+target_include_directories( KommonBase PUBLIC $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/Utility> )
 
 kasper_install_libraries( KommonBase )
 kasper_install_headers( ${KOMMON_BASE_HEADER_FILES} )

--- a/Kommon/Base/Utility/KConst.h.in
+++ b/Kommon/Base/Utility/KConst.h.in
@@ -1,0 +1,10 @@
+#if @KConst_REFERENCE_EPOCH@ == 2006
+#include "KConst_2006.h"
+
+#elif @KConst_REFERENCE_EPOCH@ == 2021
+#include "KConst_2021.h"
+
+#else
+#error "Unsupported value for KConst_REFERENCE_EPOCH."
+
+#endif

--- a/Kommon/Base/Utility/KConst.h.in
+++ b/Kommon/Base/Utility/KConst.h.in
@@ -5,6 +5,6 @@
 #include "KConst_2021.h"
 
 #else
-#error "Unsupported value for KConst_REFERENCE_EPOCH."
+#error "Unsupported value for KConst_REFERENCE_EPOCH: '@KConst_REFERENCE_EPOCH@'"
 
 #endif

--- a/Kommon/Base/Utility/KConst_2021.h
+++ b/Kommon/Base/Utility/KConst_2021.h
@@ -7,11 +7,6 @@
  * @author S. Hickford <stephanie.hickford@kit.edu>
  */
 
-#if KConst_REFERENCE_EPOCH == 2006
-#include "KConst_2006.h"
-
-#elif KConst_REFERENCE_EPOCH == 2021
-
 #ifndef KCONST_H_
 #define KCONST_H_
 
@@ -457,7 +452,3 @@ constexpr double Ue3sq()
 } /* namespace katrin */
 
 #endif  //KCONST_H
-
-#else
-#error "Unsupported value for KConst_REFERENCE_EPOCH."
-#endif

